x-pack/packages/ai-infra/inference-common/src/output/api.ts

```diff
@@ -6,39 +6,141 @@
+ *   id: 'extract_from_article',
+ *   connectorId: 'my-connector connector',
+ *   schema: mySchema,
+ *   input: `
+ *     Please find all the animals that are mentioned in the following document:
+ *     ## Document¬
+ *     ${theDoc}
+ *   `,
+ * });
+ *
+ * // output is properly typed from the provided schema
+ * const { animals } = response.output;
+ * ```
  */
 export type OutputAPI = <
   TId extends string = string,
-  TOutputSchema extends ToolSchema | undefined = ToolSchema | undefined
+  TOutputSchema extends ToolSchema | undefined = ToolSchema | undefined,
+  TStream extends boolean = false
 >(
-  id: TId,
-  options: {
-    connectorId: string;
-    system?: string;
-    input: string;
-    schema?: TOutputSchema;
-    previousMessages?: Message[];
-    functionCalling?: FunctionCallingMode;
-  }
-) => OutputResponse<TId, TOutputSchema>;
+  options: OutputOptions<TId, TOutputSchema, TStream>
+) => OutputCompositeResponse<TId, TOutputSchema, TStream>;
+
+/**
+ * Options for the {@link OutputAPI}
+ */
+export interface OutputOptions<
+  TId extends string = string,
+  TOutputSchema extends ToolSchema | undefined = ToolSchema | undefined,
+  TStream extends boolean = false
+> {
+  /**
+   * The id of the operation.
+   */
+  id: TId;
+  /**
+   * The ID of the connector to use.
+   * Must be a genAI compatible connector, or an error will be thrown.
+   */
+  connectorId: string;
+  /**
+   * Optional system message for the LLM.
+   */
+  system?: string;
+  /**
+   * The prompt for the LLM.
+   */
+  input: string;
+  /**
+   * The {@link ToolSchema} the response from the LLM should adhere to.
+   */
+  schema?: TOutputSchema;
+  /**
+   * Previous messages in the conversation.
+   * If provided, will be passed to the LLM in addition to `input`.
+   */
+  previousMessages?: Message[];
+  /**
+   * Function calling mode, defaults to "native".
+   */
+  functionCalling?: FunctionCallingMode;
+  /**
+   * Set to true to enable streaming, which will change the API response type from
+   * a single promise to an event observable.
+   *
+   * Defaults to false.
+   */
+  stream?: TStream;
+}
+
+/**
+ * Composite response type from the {@link OutputAPI},
+ * which can be either an observable or a promise depending on
+ * whether API was called with stream mode enabled or not.
+ */
+export type OutputCompositeResponse<
+  TId extends string = string,
+  TOutputSchema extends ToolSchema | undefined = ToolSchema | undefined,
+  TStream extends boolean = false
+> = TStream extends true
+  ? OutputStreamResponse<TId, TOutputSchema>
+  : Promise<
+      OutputResponse<
+        TId,
+        TOutputSchema extends ToolSchema ? FromToolSchema<TOutputSchema> : undefined
+      >
+    >;
+
+/**
+ * Response from the {@link OutputAPI} when streaming is not enabled.
+ */
+export interface OutputResponse<TId extends string = string, TOutput extends Output = Output> {
+  /**
+   * The id of the operation, as provided as input
```

Comment:
I don't love this suggestion, so feel free to disregard if you like it less than what you have.

```suggestion
   * The id of the operation, as specified by input
```